core/types/src/report.rs

```diff
@@ -128,72 +132,67 @@ new_type_accessors_impls! {
-                .try_into()
-                .expect("Failed to extract `isv_prod_id`"),
-        );
-        body.0.isv_svn = u16::from_le_bytes(
-            bytes[258..260]
-                .try_into()
-                .expect("Failed to extract `isv_svn`"),
-        );
-        body.0.config_svn = u16::from_le_bytes(
-            bytes[260..262]
-                .try_into()
-                .expect("Failed to extract `config_svn`"),
-        );
-        let isv_family_id: [u8; SGX_ISV_FAMILY_ID_SIZE] = bytes[304..320]
-            .try_into()
-            .expect("Failed to extract `isv_family_id`");
-        body.0.isv_family_id = FamilyId::from(isv_family_id).into();
-        let report_data: [u8; SGX_REPORT_DATA_SIZE] = bytes[320..384]
-            .try_into()
-            .expect("Failed to extract `report_data`");
-        body.0.report_data = ReportData::from(report_data).into();
-        body
+        let (bytes, cpu_svn) = take(SGX_CPUSVN_SIZE)(bytes)?;
+        body.0.cpu_svn = CpuSvn::try_from(cpu_svn)?.into();
+
+        let (bytes, misc_select) = le_u32(bytes)?;
+        body.0.misc_select = misc_select;
+
+        let (bytes, reserved1) = take(SGX_REPORT_BODY_RESERVED1_BYTES)(bytes)?;
+        body.0.reserved1.as_mut().copy_from_slice(reserved1);
+
+        let (bytes, extended_prod_id) = take(SGX_ISVEXT_PROD_ID_SIZE)(bytes)?;
+        body.0
+            .isv_ext_prod_id
+            .as_mut()
+            .copy_from_slice(extended_prod_id);
+
+        let (bytes, flags) = le_u64(bytes)?;
+        body.0.attributes.flags = flags;
+
+        let (bytes, xfrm) = le_u64(bytes)?;
+        body.0.attributes.xfrm = xfrm;
+
+        let (bytes, mr_enclave) = take(SGX_HASH_SIZE)(bytes)?;
+        body.0.mr_enclave = MrEnclave::try_from(mr_enclave)?.into();
+
+        let (bytes, reserved2) = take(SGX_REPORT_BODY_RESERVED2_BYTES)(bytes)?;
+        body.0.reserved2.as_mut().copy_from_slice(reserved2);
+
+        let (bytes, mr_signer) = take(SGX_HASH_SIZE)(bytes)?;
+        body.0.mr_signer = MrSigner::try_from(mr_signer)?.into();
+
+        let (bytes, reserved3) = take(SGX_REPORT_BODY_RESERVED3_BYTES)(bytes)?;
+        body.0.reserved3.as_mut().copy_from_slice(reserved3);
+
+        let (bytes, config_id) = take(SGX_CONFIGID_SIZE)(bytes)?;
+        body.0.config_id.as_mut().copy_from_slice(config_id);
+
+        let (bytes, isv_prod_id) = le_u16(bytes)?;
+        body.0.isv_prod_id = isv_prod_id;
+
+        let (bytes, isv_svn) = le_u16(bytes)?;
+        body.0.isv_svn = isv_svn;
+
+        let (bytes, config_svn) = le_u16(bytes)?;
+        body.0.config_svn = config_svn;
+
+        let (bytes, reserved4) = take(SGX_REPORT_BODY_RESERVED4_BYTES)(bytes)?;
+        body.0.reserved4.as_mut().copy_from_slice(reserved4);
+
+        let (bytes, isv_family_id) = take(SGX_ISV_FAMILY_ID_SIZE)(bytes)?;
+        body.0.isv_family_id.as_mut().copy_from_slice(isv_family_id);
+
+        let (_, report_data) = take(SGX_REPORT_DATA_SIZE)(bytes)?;
```

Comment:
Are we OK with the input hypothetically being too long? Or should it return an error? I suppose I am OK with 'too long' in this case as this is `TryFrom<&[u8]>` and therefore doesn't take the bytes, but it's a bit odd...

Reply:
Initially I was leaning toward *not* barking on too long.
The main reason being that the report body is often in the middle of a byte stream so this would force callers to end the slice.
However since this `TryFrom` doesn't consume from the slice, like nom, the callers still end up having to jump over the report body to the next item in the stream so it doesn't gain much.

In general I would say don't place an unnecessary constraint unless we can think of a reason, but I'm not fully sold here either.